### PR TITLE
[ENH] Allow README file extensions.

### LINKED ISSUE #972: Allow extension for the README file
GitLab cannot render markdown if the file doesn't have an extension. Currently, BIDS Validator complains that there is no `README` file although `README.md` is there. I think this is due to the following rule:

https://github.com/bids-standard/bids-specification/blob/febae4fb9b2c9f055dc1392e0283c85f05618de0/src/schema/rules/top_level_files.yaml#L2:L5

Could we change that and support file extensions for the `README` file? I know it would require a few changes in the documentation and maybe the validator but if we could work out the list, I'm willing to work on it.

### PR DESCRIPTION
**Pull request by mateuszpawlik**
Closes  #972

Tasks to complete:

- [x] Modify the README description in documentation https://github.com/bids-standard/bids-specification/blob/master/src/03-modality-agnostic-files.md#readme
- [x] Modify the description in the schema https://github.com/bids-standard/bids-specification/blob/ba11771e76978264445cfc35d52eda14d1ca30fc/src/schema/objects/top_level_files.yaml#L24-L26
- [x] Add new file extensions https://github.com/bids-standard/bids-specification/blob/ba11771e76978264445cfc35d52eda14d1ca30fc/src/schema/rules/top_level_files.yaml#L5
- [x] https://github.com/bids-standard/bids-validator/issues/1431 - I think the validator should be updated before the specification changes are merged in. (PR here: https://github.com/bids-standard/bids-validator/pull/1432)

I kept the README description in the specification and in the schema equal. Maybe a macro could help not to need to update such definitions in two places.

### CODE REVIEW

File: src/03-modality-agnostic-files.md
Comment (sappelhoff) [inline src/03-modality-agnostic-files.md:146]: I think we should add a short snippet of text (with links) that these extensions refer to "markdown", "restructured text", and "generic text" (or similar) respectively, and explicitly state that BIDS does not make recommendations with regards to the "flavor" (e.g. Markdown flavor) and does not validate md/rst syntax.

Instead, I think we should RECOMMEND that READMEs are written from a mindset that somebody reading the file in an un-rendered state should still be able to perfectly understand everything.

more opinions by @effigies @yarikoptic perhaps?

Comment (mateuszpawlik) [reply src/03-modality-agnostic-files.md:146]: Shall I simply implement your comments or we're waiting for more feedback?

Comment (sappelhoff) [reply src/03-modality-agnostic-files.md:146]: Both are fair approaches, the prior might result in extra work, but also potentially easier/better reviews by the reviewers :-)

Comment (tsalo) [reply src/03-modality-agnostic-files.md:146]: This kind of makes me think we should add extensions to the schema, so that the description could just link to the glossary instead of describing each of the extensions itself.

**Comment by Remi-Gau** [comment]
> I agree that we probably shouldn't update LICENSE or CHANGES given that they are more restrictive than README. If someone requests them in the future, we can revisit.

I 1000% agree with this

File: src/03-modality-agnostic-files.md
Comment (mateuszpawlik) [reply src/03-modality-agnostic-files.md:146]: I don't know what's the rule. Shall I rephrase the text now and add the links later or wait for the glossary to be ready? I could add the links now if I new how they'll be structured.

Comment (sappelhoff) [reply src/03-modality-agnostic-files.md:146]: I suggest you add a recommendations along the line of

> Instead, I think we should RECOMMEND that READMEs are written from a mindset that somebody reading the file in an un-rendered state should still be able to perfectly understand everything.

And then we can merge this PR (if other people agree) -- @tsalo (or me) can then add the links in a follow up PR (or in #1063 directly) once #1063 is merged.

File: src/schema/objects/top_level_files.yaml
Comment (yarikoptic) [inline src/schema/objects/top_level_files.yaml:28]: just to make sure - the version with absent extension (as in current BIDS) corresponds to `.txt` (unstructured text) format.  I wonder there is a value from making it explicit -- probably not, but thought to mention. WDYT?

**Review by yarikoptic — COMMENTED** [review]
mostly just non-consequential comments, with one concern -- why not to restrict .md to basic?  I bet openneuro and other portals would prefer to render those .md files. If flavor is not regulated -- it would cause troubles, thus reducing the benefit from having .md to start with.

File: src/schema/objects/top_level_files.yaml
Comment (mateuszpawlik) [reply src/schema/objects/top_level_files.yaml:24]: I also can't find a better term. It's free-form in terms of content, not format or syntax.

File: src/schema/objects/top_level_files.yaml
Comment (mateuszpawlik) [reply src/schema/objects/top_level_files.yaml:28]: From my experience, some software may not try to render a file without extension. Other than that, I personally don't see any advantage.

File: src/03-modality-agnostic-files.md
Comment (mateuszpawlik) [reply src/03-modality-agnostic-files.md:152]: I think the goal was not to validate the syntax. Therefore, no recommendation for the flavor. In my opinion basic would be too restrictive. For example tables and task lists are useful but not included in Basic syntax.

**Review by sappelhoff — APPROVED** [review]

File: src/schema/objects/top_level_files.yaml
Comment (sappelhoff) [reply src/schema/objects/top_level_files.yaml:28]: Some people may write in markdown syntax in a `README` file, so I wouldn't make it seem like the lack of an extension somehow means that the contents should be unstructured.

File: src/schema/objects/top_level_files.yaml
Comment (sappelhoff) [reply src/schema/objects/top_level_files.yaml:24]:
```suggestion
    A REQUIRED text file, `README`, SHOULD describe the dataset in more detail.
```

I suggest to remove "free-form" as it's ambiguous. A couple of sentences later, the text points to a guideline of what a good `README*` file should include.

File: src/03-modality-agnostic-files.md
Comment (sappelhoff) [reply src/03-modality-agnostic-files.md:152]: > I think the goal was not to validate the syntax.

indeed, this has been discussed in several threads and in two (or so) online meetings, notes can be found here:

- https://github.com/bids-standard/bids-validator/pull/1281#discussion_r630118909
- https://github.com/bids-standard/bids-specification/issues/972
- Maintainer Biweekly meeting notes (links to a specific comment therein): https://docs.google.com/document/d/1MuwZt6ToZmsPgH6naFVlPDFtNUz61DzlaZmdNHIAotg/edit?disco=AAAAUZdMt60

> In my opinion basic would be too restrictive. 

agreed. Regarding your concern:

1. People will write their README.md file with flavors such that it renders **at least**  on the platform(s) of their choice. For all other platforms, our recommendation kicks in: READMEs SHOULD be understandable/parsable even in unrendered form
2. READMEs will most likely not include very very advanced MD syntax, and most flavors have significant overlap, which furthermore reduces the problem

I therefore stand by the current proposal to not recommend any flavor.

**Comment by sappelhoff** [comment]
before I forget it @mateuszpawlik could you please add yourself to the contributors list here? https://github.com/bids-standard/bids-specification/wiki/Recent-Contributors

We sync that wiki entry with the corresponding page in the BIDS specification before every new release.

File: src/03-modality-agnostic-files.md
Comment (sappelhoff) [reply src/03-modality-agnostic-files.md:152]: 1. many people already use MD and have to put their README in .bidsignore to make their dataset valid
2. other people add a txt extension because they are used to having extensions, and are then surprised that this does not work
3. as I mention above: the currently proposed solution will work well for almost all cases: Those cases where people use basic markdown syntax and perhaps a markdown table ... for cases where the syntax cannot be properly rendered due to a flavor incompatibility between README and rendering host, we still have ensured the readability of the README (albeit slightly more ugly; but note: this will happen rarely)

Overall I believe that the current solution (no flavor recommendation, no validation) is a good tradeoff between utility gains for users and validation/specification burden for BIDS. 

> even in the minority I am, I still feel wrong to give up on this

Please feel free to tag more people if you want to discuss this more broadly -- I hope that we can find a consensus.

File: src/schema/objects/top_level_files.yaml
Comment (yarikoptic) [inline src/schema/objects/top_level_files.yaml:28]: it is somewhat implied but may be worth explicit something like:
```suggestion
    or `.txt`.
    Only a single README file (with or without extension) should be present.
```
?

File: src/schema/rules/top_level_files.yaml
Comment (yarikoptic) [inline src/schema/rules/top_level_files.yaml:8]: FWIW -- inspired me to file https://github.com/bids-standard/bids-specification/issues/1047

File: src/schema/objects/top_level_files.yaml
Comment (sappelhoff) [reply src/schema/objects/top_level_files.yaml:28]: agreed, though I would phrase it:

> a BIDS dataset MUST NOT contain more than one README file (with or without extension) at its root directory.

Comment (mateuszpawlik) [reply src/schema/objects/top_level_files.yaml:28]: Fixed.

**Review by sappelhoff — APPROVED** [review]

**Review by Remi-Gau — APPROVED** [review]

**Comment by sappelhoff** [comment]
Thanks for this contribution @mateuszpawlik!

**Comment by mateuszpawlik** [comment]
Thank you all for making it happen. That was a nice experience :-) I'm looking forward to more.

**Pull request merged**
